src/benchmarklib/benchmark_runner.hpp

```diff
@@ -50,15 +52,18 @@ class BenchmarkRunner {
   void _warmup_query(const QueryID query_id);
 
   // Calls _schedule_query if the scheduler is active, otherwise calls _execute_query and returns no tasks
-  std::vector<std::shared_ptr<AbstractTask>> _schedule_or_execute_query(const QueryID query_id,
-                                                                        const std::function<void()>& done_callback);
+  std::vector<std::shared_ptr<AbstractTask>> _schedule_or_execute_query(
+      const QueryID query_id, const std::shared_ptr<SQLPipelineMetrics>& metrics,
```

Comment:
Mmmh, passing in the metrics here is really not nice. Could these be a member of the class?

Reply:
We want to add them IFF and precisely when the query finished before its time was up. So we have to add them to the result in the `on_query_done` callback. I do not like it either, but haven't found a quick and clean way around it.

Reply:
The callback already has the metrics, correct? Maybe pass the pipeline as a parameter (not a capture) into the callback?

Reply:
I hope what I know changed it to is what you were looking for because the BenchmarkRunner is an otter pain to work on :(

Reply:
Argh, crap. I have only seen `BenchmarkRunner::_execute_query` and didn't realize that only there, `done_callback();` could have been extended with the pipeline as a parameter.

This new way feels a bit weird in that we are passing in the `query_id` and a pipeline which was already tailored to that particular query. Also, having three places in which we build the pipeline makes it easy to change one and forget about the others.

Please decide which way you like better - I am fine with both.

Reply:
There is now `BenchmarkRunner::_build_sql_pipeline(QueryID query_id)` that should do it.